File: java/spring-boot-mysql/src/test/java/com/meridal/examples/springbootmysql/domain/SongTest.java
package com.meridal.examples.springbootmysql.domain;

import com.meridal.examples.springbootmysql.test.TestFramework;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertFalse;

public class SongTest extends TestFramework {

        private static final Integer ID = 1234;
        private static final Integer OTHER_ID = 5678;

        @Test
        public void testEqualsAndHashCodeWithSameID() {
            final Song song = this.randomSong(ID);
            final Song other = this.cloneSong(ID, song);
            assertEquals(song, other);
            assertEquals(song.hashCode(), other.hashCode());
        }

        @Test
        public void testEqualsAndHashCodeWithDifferentID() {
            final Song song = this.randomSong(ID);
            final Song other = this.cloneSong(OTHER_ID, song);
            assertNotEquals(song, other);
            assertNotEquals(song.hashCode(), other.hashCode());
        }

        @Test
        public void testEqualsAndHashCodeWithNullID() {
            final Song song = this.randomSong();
            final Song other = this.cloneSong(ID, song);
            assertNotEquals(song, other);
            assertNotEquals(song.hashCode(), other.hashCode());
        }

        @Test
        public void testEqualsAndHashCodeWithBothIDsNull() {
            final Song song = this.randomSong();
            final Song other = this.cloneSong(null, song);
            assertNotEquals(song, other);
            assertEquals(song.hashCode(), other.hashCode());
        }

        @Test
        public void testEqualsWithAnotherType() {
            final Song song = this.randomSong();
            final List<Integer> other = new ArrayList<>();
            assertFalse(song.equals(other));
        }

        @Test
        public void testEqualsWithNull() {
            final Song song = this.randomSong();
            assertFalse(song.equals(null));
        }
}

File: kotlin/kotlin-spring-boot-mysql/src/test/kotlin/com/meridal/examples/kotlinspringboot/model/SongTest.kt
package com.meridal.examples.kotlinspringboot.model

import com.meridal.examples.kotlinspringboot.test.TestFramework
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class SongTest(): TestFramework() {

    private val ID = 1234L
    private val OTHER_ID = 5678L
    private val NOT_PERSISTED = 0L

    @Test
    fun `Songs with the same ID are equal and have the same hashCode`() {
        println("Same ID")
        val song = randomSong(ID)
        val other = cloneSong(ID, song)
        Assertions.assertEquals(song, other)
        Assertions.assertEquals(song.hashCode(), other.hashCode())
    }

    @Test
    fun `Songs with different IDs are not equal and do not have the same hashCode`() {
        val song = randomSong(ID)
        val other = cloneSong(OTHER_ID, song)
        Assertions.assertNotEquals(song, other)
        Assertions.assertNotEquals(song.hashCode(), other.hashCode())
    }

    @Test
    fun `Songs that have not been persisted are not equal`() {
        val song = randomSong(ID)
        val other = cloneSong(NOT_PERSISTED, song)
        val another = cloneSong(NOT_PERSISTED, song)
        Assertions.assertNotEquals(song, other)
        Assertions.assertNotEquals(other, another)
    }

    @Test
    fun `Songs are not equal to other types`() {
        val song = randomSong(ID)
        val list = listOf(1, 2, 3)
        Assertions.assertFalse(song == list)
    }

    @Test
    fun `A song is not equal to a null object`() {
        val song = randomSong(ID)
        val other: Song? = null
        Assertions.assertNotEquals(song, other)
    }
}